Add Field, FieldGroup, and Form components

edburyenegren-okta opened the pull request:
These new components - Field, FieldGroup, and Form - aim to abstract the duplicate code in our current input components and provide a friendlier solution for building forms.

jeffbelser-okta commented on packages/odyssey-react/src/components/Form/index.tsx:
```diff
@@ -0,0 +1,114 @@
+}
+
+type PropsFormActions = {
+  children: ReactNode
```
Being restrictive is fine, but we'll have to enforce some of this with additional code as typescript cannot enforce a specific component use via children, only generalized groups. Remember that we've also talked about enforcing useage patterns like this outside the component code itself via a lint package. Can you keep track of these types of restrictions someplace in JIRA so we can have a good batch of work to pull from for that future effort? cc @arnoldsandoval-okta

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/index.tsx:
```diff
@@ -0,0 +1,152 @@
+  );
+
+  const errorElement = (
+    <p className={styles.error} id={`${oid}-error`}>
```
got it - this is fine, but perhaps better suited as a nicely described user acceptance criteria bullet point in the JIRA that we hand off to the engineer who finalizes the implementation? I worry about having some of this nuance lost in translation!

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/index.tsx:
```diff
@@ -0,0 +1,152 @@
+
+  const labelElement = (
+    <label
+      // This should swap to legend for Radio and Checkbox Groups
```
Got it - thanks for that reminder I'm not sure why I didnt' grok that initially 😅 . Yea what we'll need is an API that allows this type of information to be injected in via a prop and have the resulting HTML change accordingly

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/Field.module.scss:
```diff
@@ -0,0 +1,55 @@
+}
+
+.hint {
+  margin-block-start: -$spacing-xs; // Hug label
```
Got it - we can leave this as is for now, but keep in mind that even component groupings that are all assigned as statics to the same base like `Form.Field` or `Form.FieldGroup` can have contextual styles. In general we want to work to avoid these types of CSS solutions because they are brittle.

A more verbose, but explicit way to solve this is to have two classes applied to label, one that is applied as a base and contains sensible defaults for margin, and a second that is applied via JS only when a hint is rendered alongside so that the margin is updated correctly on label itself without having to use negative margin on an adjacent hint element cc @arnoldsandoval-okta @andrewberg-okta

jeffbelser-okta reviewed (commented):
strong work 💪 

more thoughts to share - is this one ready for the handoff / pairing with @arnoldsandoval-okta this week? Would love to see the next steps start here

edburyenegren-okta commented on packages/odyssey-react/src/components/Radio/RadioGroup/index.tsx:
```diff
@@ -71,47 +61,48 @@ export interface Props
  * Radios appear as a ring shaped UI accompanied by a caption that allows
  * the user to choose only one option at a time.
  */
-const RadioGroup = forwardRef<HTMLFieldSetElement, Props>((props, ref) => {
```
Discussed this removal in #odyssey-core. Short version: this would require Field to have a polymorphic ref based on the associated `as` prop. Because this isn't necessary for any functionality in the RadioGroup container, we've nixed this until we have a standard pattern.

jeffbelser-okta reviewed (commented):
strong work 💪 some easy improvements to consider here!

edburyenegren-okta commented on packages/odyssey-react/src/components/Field/index.tsx:
```diff
@@ -0,0 +1,159 @@
+
+const Error = ({ id, children, errorPrefix = "Error:" }: PropsError) => (
+  <p className={styles.error} id={`${id}-error`}>
+    <ScreenReaderText>{errorPrefix}</ScreenReaderText>
```
Fixed!

jeffbelser-okta replied:
We still have a hardcoded english string on line 158 above! We can't apply english strings even as defaults

edburyenegren-okta replied:
Hm. This is how we've been handling `optionalLabel` prior to this PR, which is why we used the same approach here.

Double prefixing is a legit concern here, but docs/guidance feel like a better solve for that problem rather than allowing a case where no prefix exists by default. 

Ex: If someone is reviewing UI, "Error: Error: This is a required field" is clearly a bug, but "This is a required field." is not.

@arnoldsandoval, any thoughts here?

arnoldsandoval-okta replied:
My concern with allowing ReactNode is that it leaves room for a developer to forget the error prefix, which then creates an accessibility usability problem.

If hardcoded English strings are an issue, then we need to solve for `optionalLabel` as well, which as shred pointed out is implemented in the same fashion. I'm not sure that providing an opinionated default (in English) should be characterized as an i18n bug as we provide developers a way to update the error prefix should their products require it.

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/index.tsx:
```diff
@@ -26,6 +26,11 @@ export interface SharedFieldTypes {
+  /**
+   * Prefix to be read by SRs before announcing an error.
+   */
+  errorPrefix?: string;
```
can we avoid this specific dedicated prop here? Why not allow `error` to be a `ReactNode` so that someone can provide a mix of visible text and `<ScreenReaderText />` as necessary. Composition FTW!

arnoldsandoval-okta replied:
Updated.

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/index.tsx:
```diff
@@ -0,0 +1,169 @@
+Field.Hint = Hint;
+Field.Error = Error;
+
+export default Field;
```
we'll need to make sure any new components are using our HOC to apply their styles in test and production builds:

`export default withStyles(styles)(Field)`

jeffbelser-okta replied:
Typescript is asking you to provide the types for your statics here. These are the components you wrote in the PR here: `Field.Error` in this case. We can't hold off on this work as it represents a necessity to make our UI work at scale by handling styles and themes. It's a merge must have even if it requires some investigation and learning to make it happen. Let's work together if you need an assist.
[
Incidentally, my PR should solve this for you without the need to provide types for generics to `withStyles`](https://github.com/okta/odyssey/pull/1098) as I did for `List` previously to make this work as I'm describing above.

jeffbelser-okta commented on packages/odyssey-react/src/components/Field/Field.module.scss:
```diff
@@ -0,0 +1,55 @@
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+.root {
```
now that we've added these new shrared components and styles - can we remove any duplicated styles from the CSS modules of Checkbox/TextInput/TextArea etc?

edburyenegren-okta replied:
Yup! During our last sync on this, we were asked to not remove those styles until after this approach was reviewed and approved.

If we feel comfortable, I am more than happy to excise all of that dupe CSS now.

jeffbelser-okta commented on packages/odyssey-react/src/components/FieldGroup/index.tsx:
```diff
@@ -0,0 +1,63 @@
+
+FieldGroup.Error = Error;
+
+export default FieldGroup;
```
`export default withStyles(styles)(FieldGoup)`

jeffbelser-okta commented on packages/odyssey-react/src/components/Form/index.tsx:
```diff
@@ -0,0 +1,88 @@
+Form.Main = Main;
+Form.Actions = Actions;
+
+export default Form;
```
`export default withStyles(styles)(Form)`

edburyenegren-okta commented:
>a few more items to address! I also see a file renamed in this PR from index.ts to packages/odyssey-react/src/components/Infobox/index.tsx which doesn't need to happen because we have no JSX in that file (its a barrel file)

@jeffbelser-okta Bah, I meant to call this out up front. This file was renamed because we have a config problem somewhere along the chain. When using a `.ts` suffix, TS and Storybook both complain that the file itself cannot be found when referenced within other stories Stories (possibly elsewhere). Ex: Infobox as a child of FieldGroup.

While debugging, I noticed that Checkbox, despite undergoing the same refactor as Infobox, was not having an issue because it's `.tsx` suffix was left in place. I've renamed Infobox to unblock this PR, since that regression does not appear specific to this work.

If there's something we've implemented incorrectly in our Form or FieldGroup stories, I'd be happy to fix that and rename. Beyond that, I'd prefer to get this code merged and resolve our filetype issues separately.

jeffbelser-okta reviewed (changes requested):
This is not far from being ready to merge. Can we work together on this today to get it across the line?

We need at have wrapping `withStyles` for all our new components as part of this work. We also [need to have the `error` prefix API fixed to solve the i18n issues prior to merge.](https://github.com/okta/odyssey/pull/1050#discussion_r726194091)

But as you mention, we should immediately create two new JIRAs to follow up on the remaining parts of the job:

- Refactoring [all the existing component CSS modules to delete duplication and allow Field components](https://github.com/okta/odyssey/pull/1050#discussion_r725318469). That lets the new components deliver on their goal of DRY CSS and encapsulated component styles. Part of this work is also [removing the global form styles we're making available to the storybook environment](https://github.com/okta/odyssey/blob/develop/packages/odyssey-storybook/src/index.ts#L105-L110) here.

- Solving the `.ts|.tsx` configuration issue that [you have reported.](https://github.com/okta/odyssey/pull/1050#issuecomment-939133903)

- Fixing the duplicate or even ignored types from the `SharedFieldTypes` work you introduced in this PR. See [here](https://github.com/okta/odyssey/pull/1050/files#r725229154) and [here](https://github.com/okta/odyssey/pull/1050/files#r725228000)

edburyenegren-okta commented on packages/odyssey-react/src/components/Radio/RadioGroup/index.tsx:
```diff
@@ -75,47 +71,48 @@ export interface Props
  * the user to choose only one option at a time.
  */
-const RadioGroup = forwardRef<HTMLFieldSetElement, Props>((props, ref) => {
+const RadioGroup = (props: Props) => {
```
@jeffbelser-okta we went with your suggestion here, but the linter isn't having it:

```Missing return type on function  @typescript-eslint/explicit-module-boundary-types```

Do you have a recommendation?

jeffbelser-okta replied:
the linter is telling us something useful that its generally easier to reason about types across module boundaries if functions explicitly type their return values. In our case here you can always inspect the result of the value your function is assigned to see the inferred types, so: `JSX.Element`

all together:
```
const RadioGroup = (props: Props): JSX.Element => {
```

arnoldsandoval-okta replied:
I see. Thanks for clarifying, I will update this accordingly.

arnoldsandoval-okta commented on packages/odyssey-react/src/components/TextInput/index.tsx:
```diff
@@ -173,21 +156,25 @@ const TextInput: FunctionComponent<Props> = (props) => {
-      </div>
-    </fieldset>
+          {input}
+        </span>
```
I like it. Fixed.

jeffbelser-okta commented on packages/odyssey-react/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -52,6 +55,7 @@ const Checkbox = forwardRef<HTMLInputElement, Props>((props, ref) => {
     onChange,
     required = true,
     indeterminate = false,
+    error,
```
so `Pick` works here, but the result is not distributive across union types so if `SharedFieldTypes` even became more complicated we would introduce a subtle type failure here more than likely.

I would suggest referencing the types more directly with index access types over Pick:

```
error: SharedFieldTypes['error']

jeffbelser-okta commented on packages/odyssey-react/src/components/Radio/RadioGroup/index.tsx:
```diff
@@ -72,7 +71,7 @@ export interface Props
  * the user to choose only one option at a time.
  */
-const RadioGroup: FunctionComponent<Props> = (props) => {
+const RadioGroup = (props: Props): JSX.Element => {
```
in the future we can look at improving this return type even further via the `ReactElement` generic.

jeffbelser-okta commented on packages/odyssey-react/src/components/Checkbox/Checkbox.stories.tsx:
```diff
@@ -21,6 +21,7 @@ export default {
   args: {
     label: "Checkbox label",
     value: "checkbox_value",
+    errorPrefix: "Error:",
```
lets make a follow up JIRA to fix this bug immediately after merging this work. If you have questions lets discuss ASAP

jeffbelser-okta commented on packages/odyssey-react/src/components/Checkbox/Checkbox.tsx:
```diff
@@ -79,12 +86,15 @@ const Checkbox = forwardRef<HTMLInputElement, Props>((props, ref) => {
     internalRef.current.indeterminate = indeterminate;
   }, [indeterminate, internalRef]);
 
+  const ariaDescribedBy = useCx(error && `${oid}-error`);
```
see my previous comments about the implicit API this introduces by using this `-error` or `-hint` suffix. Lets fix this in the immediate follow up PR

jeffbelser-okta commented on packages/odyssey-react/src/components/FieldGroup/FieldGroup.module.scss:
```diff
@@ -1,5 +1,5 @@
 /*!
- * Copyright (c) 2021-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * Copyright (c) 2018-present, Okta, Inc. and/or its affiliates. All rights reserved.
```
🤔

arnoldsandoval-okta replied:
I think this is because this file was copied from the @okta/odyssey package.

jeffbelser-okta commented on packages/odyssey-react/src/components/Form/index.tsx:
```diff
@@ -0,0 +1,88 @@
+
+Form.Error = Error;
+Form.Main = Main;
+Form.Actions = Actions;
```
I want to double check here - the order of the items is not important within the form? Because using this statics based API allows developers to place these in any order they want vs a prop driven API where we control the order....

jeffbelser-okta commented on packages/odyssey-react/src/components/Select/Select.module.scss:
```diff
@@ -518,48 +518,3 @@
-  font-size: $size-body-caption;
-  font-weight: 400;
-  line-height: $spacing-m;
-}
```
\o/

jeffbelser-okta reviewed (approved):
lots of good work here! We still need to fix a few bugs and add some polish but lets take care of that in a fast follow PR

arnoldsandoval-okta commented on packages/odyssey-react/src/components/Field/Field.stories.tsx:
```diff
@@ -0,0 +1,43 @@
+  errorPrefix: "Error:",
+};
+
+export const sharedFieldArgTypes = {
```
Has been fixed.
